Library/Homebrew/extend/os/mac/formula_cellar_checks.rb

```diff
@@ -56,10 +58,42 @@ def check_python_framework_links(lib)
     EOS
   end
 
+  def check_linkage
+    return unless formula.prefix.directory?
+    keg = Keg.new(formula.prefix)
+    checker = LinkageChecker.new(keg, formula)
+    s = ""
+
+    if checker.broken_dylibs?
+      s += <<-EOS.undent
+        Broken dylib links found:
+          #{checker.broken_dylibs.to_a * "\n          "}
+
+        It is likely the installation was broken.
+        Please check formula and make sure it is built from source.
+      EOS
+    end
+
+    if checker.undeclared_deps?
+      s += <<-EOS.undent
+        Undeclared dependencies found:
+          #{checker.undeclared_deps * "\n          "}
+
+        Formulae are required to declare all linked dependencies.
+        Please check whether these dependencies are indeed needed.
+        If so, please add all missing dependencies to the formula with:
```

Comment:
> In that case the text should elaborate on how to identify if they are needed

There is no determined way to do it without any knowledge of the formula itself.

> There is also a problem here in that the brew audit output for contributors locally for these may differ significantly from the brew audit output that would be printed by CI. ...

This is actually what this PR tries to achieve that it is to expose such problem instead of offering a solution. Because the latter one is a case by case thing.